tests/test_quantum.py

```diff
@@ -54,6 +56,46 @@ def testConstructor(self):
         self.assertEqual(quantum.endTime, endTime)
         self.assertEqual(quantum.host, host)
 
+    def testAddInputsOutputs(self):
+        """Test of addPredictedInput() method.
+        """
+        quantum = Quantum(task="some.task.object", run=None)
+
+        # start with empty
+        self.assertEqual(quantum.predictedInputs, dict())
+
+        camera = "DummyCam"
+        dsName = "test_ds"
+        storageClass = StorageClass("testref_StructuredData")
+        datasetType = DatasetType(dsName, ("camera", "visit"), storageClass)
+
+        # add one ref
+        ref = DatasetRef(datasetType, dict(camera=camera, visit=42))
+        quantum.addPredictedInput(ref)
+        self.assertIn(dsName, quantum.predictedInputs)
+        self.assertEqual(len(quantum.predictedInputs[dsName]), 1)
+        # add second ref
+        ref = DatasetRef(datasetType, dict(camera=camera, visit=43))
+        quantum.addPredictedInput(ref)
+        self.assertEqual(len(quantum.predictedInputs[dsName]), 2)
```

Comment:
Instead of checking for the lengths you may want to do `self.assertEqual(set(predicted), set(quantum.predictedInputs))`.

Reply:
Same below.

Reply:
https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertCountEqual

Reply:
Well, I though about it (maybe not enough) and I wanted to have more relaxed test because I prefer not to know what `addPredictedInput()` stores. Anyways I tried to use `assertCountEqual` and here is what I get:
```
[salnikov@centos-dev-vm daf_butler]$ python -m unittest tests/test_quantum.py 
E...
======================================================================
ERROR: testAddInputsOutputs (tests.test_quantum.QuantumTestCase)
Test of addPredictedInput() method.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/salnikov/lsstsw/lsstsw/miniconda/lib/python3.6/unittest/case.py", line 1166, in assertCountEqual
    first = collections.Counter(first_seq)
  File "/home/salnikov/lsstsw/lsstsw/miniconda/lib/python3.6/collections/__init__.py", line 537, in __init__
    self.update(*args, **kwds)
  File "/home/salnikov/lsstsw/lsstsw/miniconda/lib/python3.6/collections/__init__.py", line 624, in update
    _count_elements(self, iterable)
TypeError: unhashable type: 'DatasetRef'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/salnikov/DM-14554/daf_butler/tests/test_quantum.py", line 80, in testAddInputsOutputs
    self.assertCountEqual(quantum.predictedInputs[datasetTypeName], [ref1, ref2])
  File "/home/salnikov/lsstsw/lsstsw/miniconda/lib/python3.6/unittest/case.py", line 1170, in assertCountEqual
    differences = _count_diff_all_purpose(first_seq, second_seq)
  File "/home/salnikov/lsstsw/lsstsw/miniconda/lib/python3.6/unittest/util.py", line 137, in _count_diff_all_purpose
    if other_elem == elem:
  File "/home/salnikov/DM-14554/daf_butler/python/lsst/daf/butler/core/utils.py", line 90, in slotValuesAreEqual
    return all((getattr(self, slot) == getattr(other, slot) for slot in allSlots(self)))
  File "/home/salnikov/DM-14554/daf_butler/python/lsst/daf/butler/core/utils.py", line 90, in <genexpr>
    return all((getattr(self, slot) == getattr(other, slot) for slot in allSlots(self)))
AttributeError: 'object' object has no attribute '_id'

----------------------------------------------------------------------
Ran 4 tests in 0.031s

FAILED (errors=1)
```
(pytest output is more verbose, this is why unittests but pytest fails in the same way.

Using `assertEqual(set(), set())` I get similar but shorter traceback:
```
[salnikov@centos-dev-vm daf_butler]$ python -m unittest tests/test_quantum.py 
E...
======================================================================
ERROR: testAddInputsOutputs (tests.test_quantum.QuantumTestCase)
Test of addPredictedInput() method.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/salnikov/DM-14554/daf_butler/tests/test_quantum.py", line 76, in testAddInputsOutputs
    self.assertEqual(set(quantum.predictedInputs[datasetTypeName]), set([ref1]))
TypeError: unhashable type: 'DatasetRef'

----------------------------------------------------------------------
Ran 4 tests in 0.019s

FAILED (errors=1)
```

I'll leave it as it is now, I guess this can be improved after DatasetRef becomes hashable.